### ISSUE TITLE
Expand invoice detail to show fees/discounts/credits

### ISSUE THREAD

**Issue by nosegraze** [restrictcontentpro/restrict-content-pro#1150]
After #1000 I'd love to see more detail put into the invoices.

In one of our meetings we talked about the possibility of adding more payment columns for all the possible discounts/fees, and the idea is really growing on me. That could tie in nicely to expanding the invoices. If we added database columns for like `proration_credits`, `fees`, and maybe `discounted_amount`, we could include that kind of information on the invoices.

I got a ticket from a customer who had someone signup for a £40 plan with a discount code and the invoice just showed £20. He requested that the original £40 amount be shown, along with the amount that was discounted.

## Current
![rcp-invoice-current](https://cloud.githubusercontent.com/assets/6324272/24605540/22586874-1861-11e7-8f46-2ffe07e75a98.png)

## Proposed
![rcp-invoice-proposed](https://cloud.githubusercontent.com/assets/6324272/24605544/2918f96c-1861-11e7-965b-463aeef56bc2.png)

**Comment by tnorthcutt** [comment]
Since #1000 is closed, this can be done now. I'll take a crack at it for 3.0.

**Comment by tnorthcutt** [comment]
My WIP output looks like this:

![invoice](https://user-images.githubusercontent.com/796639/38097348-5eba9c00-3332-11e8-84dd-84d041f9dda8.png)

**Important change to note:** currently, the `Invoice Items` box includes the subscription level, and `$rcp_payment->amount`. So in the case above, that would actually show `$28`. But that's super confusing if we add in info about fees, credits, and discounts. The math ends up not making any sense, because it'd show this instead:

![invoice2](https://user-images.githubusercontent.com/796639/38097484-c12f1e4c-3332-11e8-8fd0-a440b6886931.png)

...which obviously doesn't make sense.

So, this changes things up where, at least in my mind, they make way more sense. Open to input on that, of course 😁

**Comment by nosegraze** [comment]
I think what you've done makes sense, because the table layout implies there can be multiple line items, and if that were the case it would be laid out exactly as you've done.

However, this makes me wonder if we should rethink more about how the invoice looks. As I said, it's laid out as if you can make multiple purchases (think EDD). The table is literally labeled, _"Invoice Items"_. But actually you can never buy more than one at a time. Should we think about a bigger redesign that clarifies that?

IDK, just thinking out loud.

**Comment by tnorthcutt** [comment]
Yeah, I know what you mean. Will that change once multiple subscriptions are possible? In that case, could someone pay for two subscriptions at the same time? If so, this would be set up correctly ahead of time (though a `payment` could then potentially have more than one discount, fee, etc... A whole new can of worms!).

**Comment by nosegraze** [comment]
We don't have any plans to build a whole shopping cart functionality at this time. I think it would still be one purchase at the time for the foreseeable future (hopefully forever :D).

**Comment by tnorthcutt** [comment]
@nosegraze Cool, I'll make some tweaks then.

**Comment by tnorthcutt** [comment]
Actually I think just changing `INVOICE ITEMS:` to `SUBSCRIPTION LEVEL:` or `SUBSCRIPTION:` works.

Sound ok?

**Comment by mindctrl** [comment]
Something else to consider... we talked about having the ability in the future, via an add-on, to purchase individual articles. 😆

**Comment by tnorthcutt** [comment]
🙃

**Comment by luke-au** [comment]
Will this new invoice itemise fees?

Plugin Name: Restrict Content Pro - Registration Fee
library/rcp-registration-fee.php

```php
/**
 * Adds a custom fee to registration
 *
 * This example adds a $20 fee that applies to the first payment only and is not affected by pro-ration
 *
 * @param RCP_Registration $registration
 *
 * @return void
 */
function pw_rcp_add_registration_fee( $registration ) {
	
	$registration->add_fee( 20, __( 'Custom Registration Fee', 'rcp' ), false, false );
}
add_action( 'rcp_registration_init', 'pw_rcp_add_registration_fee', 20 );
```

**Comment by nosegraze** [comment]
@luke-au I'm afraid not - sorry. We don't store each fee individually so it's impossible for us to differentiate them after registration.

**Comment by tnorthcutt** [comment]
@nosegraze does this sound ok to you?

> Actually I think just changing INVOICE ITEMS: to SUBSCRIPTION LEVEL: or SUBSCRIPTION: works.

**Comment by nosegraze** [comment]
How about this?

![rcp-invoice](https://user-images.githubusercontent.com/6324272/38240698-39a8c136-3728-11e8-8de6-6fa4c7c469fc.png)

These would be added in a `<thead>` tag.

**Comment by mindctrl** [comment]
@nosegraze I like that!

**Comment by mindctrl** [comment]
Prior to #1000 landing, we did not store subtotal, credits, fees, and discount amounts in the payments table. Ref: https://github.com/restrictcontentpro/restrict-content-pro/pull/1127/files#diff-0e76810e485377b380ecabaec99134ab

With this PR, when viewing receipts for payments without that data, you'll get a bunch some PHP notices and a funny looking invoice.

PHP notice:
`A non-numeric value encountered in /app/public/wp-content/plugins/restrict-content-pro/templates/invoice.php on line 293`

Screenshot:
![image](https://user-images.githubusercontent.com/1254870/38267759-64d78474-374a-11e8-832e-b256688acf89.png)

We can update `RCP_Payments::backfill_payment_data()` to address this. If those columns are empty, we can give them a `0` value, except the `subtotal` column, which I think should equal the `amount` column since we weren't breaking down the amounts at that point.

**Comment by tnorthcutt** [comment]
@mindctrl very nice catch 👏. Updated PR incoming.

**Comment by tnorthcutt** [comment]
Something that just occurred to me: is there a danger that some site owners will have customers who see the invoice and go "Wait a minute, I could have had a discount?!?! I want a discount!!!".

If so, they'd probably want an option to exclude any zero amounts (so fees, credits, and discounts wouldn't get shown if they have zero values).

Major downside: that's Yet Another Setting™ to add. Perhaps adding that should wait for user feedback requesting it?

**Comment by nosegraze** [comment]
Yes I'm happy with omitting rows that are `0`.

**Comment by tnorthcutt** [comment]
@mindctrl @nosegraze good call; done now

**Comment by mindctrl** [comment]
I think we should show credits and discounts as negative numbers, like we do negative fees. See these two screenshots for why it looks confusing:

![image](https://user-images.githubusercontent.com/1254870/38510975-9ea55ae2-3bf4-11e8-9817-acef93236cee.png)

![image](https://user-images.githubusercontent.com/1254870/38510955-8d4681e0-3bf4-11e8-9f69-d1b6a2c4e77b.png)

**Comment by mindctrl** [comment]
Negative fees show as negative because we store them that way. But we don't do that for discounts and credits. Whether we should store them as negatives is a different discussion, maybe? But I'd like for these numbers to be easily understandable either way.

**Comment by nosegraze** [comment]
I like that idea!

**Comment by tnorthcutt** [comment]
@mindctrl yep, makes sense to me. Changed.
